mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -761,39 +778,119 @@ propagateVectorizationInfo(TransformMapAttr map, const VectorizationData &input,
+                 << *user << "\n");
+      return failure();
+    }
+
+    if (newTransformed) {
+      LLVM_DEBUG(llvm::dbgs() << "[vectorization] Found multiple users that "
+                                 "could be the next one\n");
+      return failure();
+    }
+    newTransformed = candidate;
+    newRoot = user;
+  }
+  if (!newTransformed) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "[vectorization] This memref.alloc is a dead end\n");
+    return failure();
+  }
+  return std::make_pair(newTransformed, newRoot);
+}
+
+VectorizationResult mlir::rock::getMaxVectorization(
+    Value transformed, uint32_t dim, std::optional<int64_t> inputDimLen,
+    Operation *operationRootForFusionTraversal, bool ignoreDataType) {
+  auto upperType = cast<ShapedType>(transformed.getType());
+  int64_t numInitialDims = upperType.getRank();
+  int64_t initialVecLen = inputDimLen.value_or(upperType.getShape()[dim]);
   VectorizationData data;
-  auto contiguousMerges = findContiguousGroups(transforms, outputShape);
   // grow() takes the last index, not the length
   data.grow(numInitialDims);
-  data[dim] = VectorizationInfo(/*maxLength=*/len, /*needsCoefficient=*/1,
-                                /*alignment=*/len);
-  for (auto transformMap : transforms.getAsRange<TransformMapAttr>()) {
-    LLVM_DEBUG(llvm::dbgs() << "Max vectorization data: ");
-    data.debugPrint();
-    LLVM_DEBUG(llvm::dbgs() << "Processing: " << transformMap << "\n");
-    data = propagateVectorizationInfo(transformMap, data, contiguousMerges);
+  data[dim] =
+      VectorizationInfo(/*maxLength=*/initialVecLen, /*needsCoefficient=*/1,
+                        /*alignment=*/initialVecLen);
+  bool traverseFusions = (operationRootForFusionTraversal != nullptr);
+  Operation *currentUser = operationRootForFusionTraversal;
+  Value currentVal = transformed;
+  LogicalResult fusionTraversalStatus = success();
+
+  // In the outer loop, we handle cases that make us jump to a different
+  // transform stack. For example, if we're traversing fusions and we reach
+  // a memref.alloc, we know we need to find the fusion that reads from
+  // it in order to visit the transforms on its output. Similarly, if we see
+  // a rock.scalarize, we'll need to adjust the vectorization data
+  // to account for the underlying data type and then continue.
+  while (true) {
+    auto contiguousMerges = findContiguousGroups(transformed);
+    while (auto trOp = currentVal.getDefiningOp<TransformOp>()) {
+      TransformMapAttr transformMap = trOp.getTransform();
+      LLVM_DEBUG(llvm::dbgs() << "Max vectorization data: ");
+      data.debugPrint();
+      LLVM_DEBUG(llvm::dbgs() << "Processing: " << transformMap << "\n");
+      data = propagateVectorizationInfo(transformMap, data, contiguousMerges);
+      currentVal = trOp.getInput();
+      currentUser = trOp.getOperation();
+    }
+    if (traverseFusions && currentVal.getDefiningOp<memref::AllocOp>()) {
```

Comment:
[not a blocking comment] i think its much cleaner to write a wrapper function that returns currentVal and currentUser.

A function that would do :
So if its a TransformOp --> then return { trOp.getInput(), trOp.getOperation()}
If not && traverseFusions --> return findPostFusionTransforms(...)
(also can pass the fusionTraversalStatus get an update)

So you iterate until that wrapper function return success..

Reply:
and get rid of this `while(true)`

Reply:
The one tricky thing is that the transitions can sometimes involve affecting some other variables, but yeah, it could be a lambda.